MessagePack version

jwoodrow opened the pull request:
MessagePack is more compact and slightly faster on average than `JSON.parse`/`JSON.pretty_generate`.

In my use case I had
- 5561 tests
- The cache was 66MB when using the JSON version
- Loading the cache and rerunning the pending tests I setup took ~4 minutes
- The cache generation took a fair bit of time (I did not time this unfortunately as the whole test suite takes upwards of an hour to run)

When using MessagePack
- The cache was reduced to 34MB
- Loading the cache and rerunning the pending tests I setup took ~3 minutes
- Cache generation took under 30 seconds

Did not quite know how I could add this as a configuration option but if you could point me in the right direction I was thinking of `cache_serializer` and have it be either `JSON` or `MessagePack` basically 🤔 

-----------------

Before submitting the PR make sure the following are checked:

* [x] Wrote [good commit messages](https://chris.beams.io/posts/git-commit/).
* [x] Feature branch is up-to-date with the `main` branch.
* [x] Squashed related commits together.
* [ ] Added tests. (Don't know which ones to add yet since i's more of a replacement)
* [x] Run `bundle exec rake`.

avmnu-sng commented:
@jwoodrow You can update the `configuration.rb` file and create a method `cache_serializer` and have the default set to `message_pack`

jwoodrow commented:
@avmnu-sng

I've added an option to change the serialised and even create new custom ones (I tried my hand at doing one using `Marshal` but that one needed some pretty heavy custom code that actually makes it so that `Marshal` is basically as slow if not slower than `JSON` (even though it should be much faster)

I've also updated the readme with explanation on how it works and how to extend the Serializer to create even more custom serialisers.

I'm failing to understand how to run the full test suite locally and I think I might of incorrectly changed the specs to use the configuration but I'm not sure how I could change the `feature` files to test all 3 serialisers individually (also because I'd need to generate the matching cache files to do so). Could you perhaps add a section in the `CONTRIBUTING.md` to explain how to fully run the test suite ? I'd be happy to add the extra specs to test my implementations

avmnu-sng commented:
@jwoodrow The tests are run against Ruby 2.5, 2.6, 2.7, and 3.0. I have updated the rake task to also run feature specs so that you can see what's failing in the local development environment. Also, writing tests would be cumbersome at the moment, because I mostly have feature specs right now, and I am yet to add unit tests.

I released **[v0.6.0](https://github.com/avmnu-sng/rspec-tracer/releases/tag/v0.6.0)**. You should try it out to see which stage is taking more time and see if adding the custom serializer support is worth the effort.

jwoodrow commented:
> The tests are run against Ruby 2.5, 2.6, 2.7, and 3.0. I have updated the rake task to also run feature specs so that you can see what's failing in the local development environment. Also, writing tests would be cumbersome at the moment, because I mostly have feature specs right now, and I am yet to add unit tests.

The problem I'm having is that the `ruby_app` part of the specs are failing on my fork and I don't quite understand why or how this might be cause by my changes that by default should have identical behaviours 🤔 
 
> I released **[v0.6.0](https://github.com/avmnu-sng/rspec-tracer/releases/tag/v0.6.0)**. You should try it out to see which stage is taking more time and see if adding the custom serializer support is worth the effort.

I've already notices a 3 minute decrease in re-run time when using MessagePack > JSON for the serializing (and this is after the fix to exclude vendor from the coverage which reduced the cache size to 66M and MessagePack's version is still smaller at 34M) so I might stick with my fork for now just because 3 minutes is a lot of time save on travis. If runs usually take 20 minutes to run using spec-tracer then making it take 17 minutes can allow an extra run to finish every 1.5 hours.

I'll give you the numbers though once the run finishes.

jwoodrow commented:
I started with the MessagePack times (since I already had a cache for those

```bash
RSpec tracer loaded cache from __ (took 0.26647 seconds)
RSpec tracer processed cache (took 0.04382 seconds)
RSpec tracer is generating reports
RSpec tracer processed dependency (took 0.06904 seconds)
RSpec tracer loaded cached examples coverage (took 0.34857 seconds)
RSpec tracer processed coverage (took 2.00 seconds)
RSpec tracer generated flaky, failed, and pending examples report (took 0.00005 seconds)
RSpec tracer generated all files report (took 0.00180 seconds)
RSpec tracer generated all examples report (took 0.00553 seconds)
RSpec tracer generated dependency report (took 0.10171 seconds)
RSpec tracer generated examples coverage report (took 0.09890 seconds)
RSpec tracer generated reverse dependency report (took 0.19905 seconds)
RSpec tracer reports written to __ (took 1.22 seconds)
RSpecTracer generated HTML report to __ (took 0.08948 seconds)
```

I'll update with the JSON times tomorrow since a single run takes over an hour and a half to run and then I'd need to run it again to get the cached logs

Once thing to do might be to make the times be an optional config something like `VERBOSE` or `DEBUG` maybe 🤔

jwoodrow commented:
@avmnu-sng I had to fix your changes in 0.6.0 because you had the following code in the wrong order

```ruby
seconds, remainder = seconds.divmod(count) # remainder is a Number
remainder = format_duration(remainder) # remainder becomes a String

next if remainder.zero? # String does not have a zero? method
```

which should actually be

```ruby
seconds, remainder = seconds.divmod(count)
next if remainder.zero?

remainder = format_duration(remainder)
```

jwoodrow commented:
@avmnu-sng I'm rerunning the initial empty cache version for my MessagePack version to see if there what the difference on the first run is too. It should be noted that a single run does exactly count as a benchmark but here are the results (I'll keep them updated once the initial run for MessagePack with timers is done.

```ruby
MessagePack Version

RSpec tracer loaded cache from __ (took 0.26647 seconds)
RSpec tracer processed cache (took 0.04382 seconds)
RSpec tracer is generating reports
RSpec tracer processed dependency (took 0.06904 seconds)
RSpec tracer loaded cached examples coverage (took 0.34857 seconds)
RSpec tracer processed coverage (took 2.00 seconds)
RSpec tracer generated flaky, failed, and pending examples report (took 0.00005 seconds)
RSpec tracer generated all files report (took 0.00180 seconds)
RSpec tracer generated all examples report (took 0.00553 seconds)
RSpec tracer generated dependency report (took 0.10171 seconds)
RSpec tracer generated examples coverage report (took 0.09890 seconds)
RSpec tracer generated reverse dependency report (took 0.19905 seconds)
RSpec tracer reports written to __ (took 1.22 seconds)
RSpecTracer generated HTML report to __ (took 0.08948 seconds)

total_time = 5.24974 seconds

JSON version

1st run
RSpec tracer is running 5568 examples (actual: 5568, skipped: 0) (took 0.54421 seconds)
RSpec tracer is generating reports
RSpec tracer processed dependency (took 3.00 minutes 51.62 seconds)
RSpec tracer processed coverage (took 3.21 seconds)
RSpec tracer generated flaky, failed, and pending examples report (took 0.00002 seconds)
RSpec tracer generated all files report (took 0.00002 seconds)
RSpec tracer generated all examples report (took 0.00001 seconds)
RSpec tracer generated dependency report (took 0.00657 seconds)
RSpec tracer generated examples coverage report (took 0.00001 seconds)
RSpec tracer generated reverse dependency report (took 0.20508 seconds)
RSpec tracer reports written to __ (took 1.32 seconds)
RSpecTracer generated HTML report to __ (took 0.17147 seconds)

total_time = 237.07739 seconds (3 minutes 57.07739 seconds)

2nd run
RSpec tracer loaded cache from __ (took 0.33034 seconds)
RSpec tracer processed cache (took 0.04293 seconds)
RSpec tracer is generating reports
RSpec tracer processed dependency (took 0.09334 seconds)
RSpec tracer loaded cached examples coverage (took 1.47 seconds)
RSpec tracer processed coverage (took 2.68 seconds)
RSpec tracer generated flaky, failed, and pending examples report (took 0.00005 seconds)
RSpec tracer generated all files report (took 0.00124 seconds)
RSpec tracer generated all examples report (took 0.00491 seconds)
RSpec tracer generated dependency report (took 0.09911 seconds)
RSpec tracer generated examples coverage report (took 0.11976 seconds)
RSpec tracer generated reverse dependency report (took 0.20627 seconds)
RSpec tracer reports written to __ (took 0.53843 seconds)
RSpecTracer generated HTML report to __ (took 0.13690 seconds)

total_time = 5.72328 seconds
```

avmnu-sng commented:
> Once thing to do might be to make the times be an optional config something like VERBOSE or DEBUG maybe

@jwoodrow adding an environment variable to control this. Making some improvements on the time-formatter as well.

avmnu-sng commented:
> I'm rerunning the initial empty cache version for my MessagePack version to see if there what the difference on the first run is too. It should be noted that a single run does exactly count as a benchmark but here are the results (I'll keep them updated once the initial run for MessagePack with timers is done.

@jwoodrow The time (serialization vs deserialization) is from `loaded cache` and `reports written to __`. Let's see what improvements we are getting.

avmnu-sng commented:
@jwoodrow I don't think that the time-consuming step is serialization and deserialization. If you notice, most time is spent post-execution during dependency processing `RSpec tracer processed dependency`.
